pkg/minibroker/provider.go

```diff
@@ -17,53 +17,129 @@ limitations under the License.
+	"strings"
 
 	corev1 "k8s.io/api/core/v1"
 )
 
+// Provider is the interface for the Service Provider. Its methods wrap service-specific logic.
 type Provider interface {
-	Bind(service []corev1.Service, params map[string]interface{}, chartSecrets map[string]interface{}) (*Credentials, error)
+	Bind(
+		service []corev1.Service,
+		bindParams *BindParams,
+		provisionParams *ProvisionParams,
+		chartSecrets Object,
+	) (Object, error)
 }
 
-type Credentials struct {
-	Protocol string
-	URI      string `json:"uri,omitempty"`
-	Username string `json:"username,omitempty"`
-	Password string `json:"password,omitempty"`
-	Host     string `json:"host,omitempty"`
-	Port     int32  `json:"port,omitempty"`
-	Database string `json:"database,omitempty"`
+// Object is a wrapper around map[string]interface{} that implements methods for helping with
+// digging and type asserting.
+type Object map[string]interface{}
+
+var (
+	// ErrDigNotFound is the error for a key not found in the Object.
+	ErrDigNotFound = fmt.Errorf("key not found")
+	// ErrDigNotString is the error for a key that is not a string.
+	ErrDigNotString = fmt.Errorf("key is not a string")
+)
+
+// Dig digs the Object based on the provided key.
+// key must be in the format "foo.bar.baz". Each segment represents a level in the Object.
+func (o Object) Dig(key string) (interface{}, bool) {
+	keyParts := strings.Split(key, ".")
+	var part interface{} = o
+	var ok bool
+	for _, keyPart := range keyParts {
+		switch p := part.(type) {
+		case map[string]interface{}:
+			if part, ok = p[keyPart]; !ok {
+				return nil, false
+			}
+		case Object:
+			if part, ok = p[keyPart]; !ok {
+				return nil, false
+			}
+		default:
+			return nil, false
+		}
+	}
+	return part, ok
+}
+
+// DigString wraps Object.Dig and type-asserts the found key.
+func (o Object) DigString(key string) (string, error) {
+	val, ok := o.Dig(key)
+	if !ok {
+		return "", ErrDigNotFound
+	}
+	valStr, ok := val.(string)
+	if !ok {
+		return "", ErrDigNotString
+	}
+	return valStr, nil
 }
 
-// ToMap converts the credentials into the OSB API credentials response
-// see https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#device-object
-// {
-//   "credentials": {
-//     "uri": "mysql://mysqluser:pass@mysqlhost:3306/dbname",
-//     "username": "mysqluser",
-//     "password": "pass",
-//     "host": "mysqlhost",
-//     "port": 3306,
-//     "database": "dbname"
-//     }
-// }
-func (c Credentials) ToMap() map[string]interface{} {
-	var result map[string]interface{}
-	j, _ := json.Marshal(c)
-	json.Unmarshal(j, &result)
-	return result
+// DigStringAlt digs for any of the given keys, returning the first found. It returns an error if
+// none of the alternative keys are found.
+func (o Object) DigStringAlt(altKeys []string) (string, error) {
+	for _, altKey := range altKeys {
+		valStr, err := o.DigString(altKey)
+		if err != nil {
+			if err == ErrDigNotFound {
+				continue
+			}
+			return "", err
+		}
```

Comment:
```suggestion
		if err == ErrDigNotFound {
			continue
		}
		if err != nil {
			return "", err
		}
```
No need to have extra nesting.

Reply:
I did that on purpose to keep the same pattern of error verification. But now that I say it out loud, it sounds a bit silly.

Reply:
Fixed with d350950b3388a87b397b8530f67525b24d362512.